muck.1
.TH MUCK "1" "December 2021"
.schar \[u2026] ...
.SH NAME
muck \- music player
.
.SH SYNOPSIS
.
.RB muck
.RI  [ OPTION ]...
.RI  [ URL ]...
.
.SH DESCRIPTION
.B muck
is a terminal music player built on top of FFmpeg.
.
.PP
.B master
playlist is populated with the given
.IR URL s.
When no URLs are given and standard input does not refer to a terminal, a
special
.B stdin
playlist is created and playlist file is read from the standard input.
Otherwise a
.B \.
playlist is created and files are read from the current directory.
.
.SS Playlists
.PP
URLs ending in
.RB {\  .m3u " | " .m3u8 " | " .pl " }[ " .bz " | " .bz2 " | " .gz " | " .lz4 " | " .xz " | " .zst \ ]
are treated as (compressed) playlist files.
.PP
Playlist files that start with
.BR #EXTM3U ,
are recognized as M3U playlists and for every media URL metadata is
written to and read from the
.BR #EXTINF \ directive.
Playlists containing unknown comments are marked read-only thus never
overwritten.
.PP
.BI #PLAYLIST: NAME
directive can be used to give a pretty name for a playlist (instead of
referencing to it by its path).
.PP
Playlists are opened once at startup and may be nested without limits.
.PP
Directories are treated like a virtual playlist with the limitation that they
are recognized on the command-line only, i.e. they are neither scanned
recursively nor can be played from inside a playlist. If you would like to
overcome this limitation, use
.BR "find -type f | muck" .
.
.SS Cover art
Cover art is written to
.RB $ MUCK_COVER .
If not set, a suitable location is chosen.
.
.SH OPTIONS
.TP
.BI \-q\  QUERY
Execute initial search query.
.
.TP
.BI \-e\  COMMAND
Execute
.I COMMAND
at startup. Default:
.BR 0G " (begin playing from the first file)."
.
.TP
.BI \-a\  FILTERGRAPH
Specify filtergraph to use. See
.BR ffplay (1)
.BR -af .
.IP
Default:
.BR volume=replaygain=track .
Use
.B anull
to disable.
.
.TP
.BI \-c\  CODEC
Set output codec. Default:
.BR pcm " (fake codec that uses sample format of input)."
.
.TP
.BI \-f\  FORMAT
Set output format. Default:
.BR alsa .
.
.TP
.BI \-o\  FILENAME
Set output filename. Default: (none).
.
.TP
.BI \-m\  SIZE
Set maximum demuxer buffer size in kB. Default:
.BR 4096 .
.
.TP
.BI \-C\  SPEC
Specify column layout. Default:
.BR iy30a,x25A+Fd*20Tn*40t+f+vlgbIB*LCoOm*z .
.IP
.IR SPEC " := [ [" MODIFIER "] [" WIDTH "] " KEY " ]..."
.IP
.TS
tab(|)[allbox];
lil
ll
lbl.
MODIFIER|Description
\[u2423]|Join with " ".
*|Make column flexible. \fIWIDTH\fR specifies minimum width.
+|Wrap in " (\[u2026])".
,|Join with ";".
-|Join with " - ".
/|Join with " / ".
.TE
.IP
.I WIDTH
specifies required column width. When omitted a default value is used. As a
special case, 0 means key should not be columnized but join it to the previous
column.
.IP
.I KEY
is a single letter metadata key. The complete list of recognized keys is shown
in the search screen.
.
.TP
.BI \-v
Print version information and exit.
.
.SH KEY BINDINGS
.SS Live mode
Live mode is the default.
.TP
.BR 0\-9
Set
.IR N
for next command.
.
.TP
.BR p ,\  N ,\  Up
Seek to
.IR N "th (default: 1) previous file."
.
.TP
.BR n ,\  Down
Seek to
.IR N "th (default: 1) next file."
.
.TP
.BR g ,\  Home
Seek to
.IR N "=MMSS (default: 0)."
.IP
For example
.BR 100g ,\  60g
both seek to 1 minute.
.
.TP
.BR G ,\  End
Seek to
.IR N "th (default: last) file"
.
.TP
.BR r
Seek to a random file (other than the currently playing
file).
.
.TP
.BR Ctrl+f ,\  Ctrl+b ,\  PageDown ,\  PageUp
Move selection by pages.
.
.TP
.BR Ctrl+d ,\  Ctrl+u
Move selection by half pages.
.
.TP
.BR Return
Seek again. Seek command to be repeated is displayed at
the left of the status line.
.
.TP
.BR e ,\  b
Seek by 1 and -2 seconds, respectively.
.
.TP
.BR h ,\  l ,\  Left ,\  Right
Seek by 5 seconds.
.
.TP
.BR j ,\  k
Seek by 5% but at least 5 seconds.
.
.TP
.BR *
Set volume to
.I N
if given, otherwise toggle mute. Uses
.B volume
filter.
.
.TP
.BR + ,\  \-
Change
.B volume
property of
.B volume
filter by +1% and -2%, respectively.
.
.TP
.BR t
Switch audio track. When multiple tracks are available it
is shown in the status line (rare).
.
.TP
.BR .
Pause playback.
.
.TP
.BR >
Unpause playback.
.
.TP
.BR Space ,\  c
Toggle pause playback.
.
.TP
.BR f ,\  /
Search files. See "SEARCHING".
.
.TP
.BR F
Reset search filter.
.
.TP
.BR A
Search for "a", i.e. get songs from current artist.
.
.TP
.BR B
Search for "a A", i.e. get current album.
.
.TP
.BR T
Search for "a t", i.e. get versions of current song.
.
.TP
.BR o ,\  =
Sort files. See "SORTING".
.
.TP
.BR O
Disable sorting (empty sort specifier).
.
.TP
.BR I
Sort by "i=p".
.
.TP
.BR R
Sort by "aTtn=".
.
.TP
.BR J ,\  K
Move file by 1 step or move to
.IR N "th position"
when given.
.
.TP
.B |
Print files to standard output.
.
.TP
.BR H ,\  L ,\  Shift+Left ,\  Shift+Right
Scroll file list horizontally.
.
.TP
.B w
Toggle column widening. Ignore narrow columns (\[<=]10 cells).
.
.TP
.B v
Switch between live and visual modes.
.
.TP
.BR i
Toggle show stream information.
.
.TP
.BR Ctrl+L
Clear error message, if any, otherwise refresh screen.
.
.TP
.BR ? ,\  F1
Open manual.
.
.TP
.BR q ,\  Z
Quit.
.
.TP
(other letter)
Execute
.RB $ MUCK_HOME/{KEY} .
.IP
.RB $ MUCK_ *
environment variables are set, based on the currently playing file.
.
.SS Visual mode
Key bindings are mostly the same as in live mode, except
selection does not start playing after seeking.
.
.TP
.BR Return
Start playing current file.
.
.TP
.BR j ,\  k
Move selection by
.IR N " (default: 1) steps."
.
.TP
.BR g ,\  Home
Go to top.
.
.SH SEARCHING
A temporary file is opened and shown for edit. All lines but the first one is
ignored.
.P
.I FIRST-LINE
:=
.I EXPR
.br
.
.I EXPR
:=
.RI [ KEY ]...
[ "?" ] [ { "<" | ">" }[ "=" ] | "~" ]
.RI [ VALUE ]
.
.br
.I EXPR
:= "!"
.I EXPR
.
.br
.I EXPR
:=
.I EXPR
"&"
.I EXPR
|
.I EXPR
.I EXPR
.
.br
.I EXPR
:=
.I EXPR
"|"
.I EXPR
.
.br
.I VALUE
:=
.I QUOTED
|
.I WORD
.
.br
.I QUOTED
:= "'" [ all characters - "'" ]... "'"
.
.br
.I QUOTED
:= '"' [ all characters - '"' ]... '"'
.
.br
.I WORD
:= { all characters - "'", '"', " ", "|", ")" } [ all characters - " ", "|", ")" ]...
.
.P
Tokens may be separated by " " (space).
.
.P
If
.I KEY
is omitted, it defaults to keys marked with "+". (The list is shown in the
search screen, though it is rarely interesting.) In practice, it means almost
all known metadata, in order to ease search of unscanned files.
.IP
.TS
tab(@)[];
lbl
lb^
lb^.
~love.*bugs@Searches artist, title, url, comment...
'all star'
<=1001
.TE
.
.P
When multiple
.IR KEY s
are specified, it matches when any of them is matching.
.IP
.TS
tab(@)[];
lbl.
axf^Don@Same as "a^Don | x^Don | f^Don".
.TE
.
.TP
.BR < ,\  > ,\  =
Compare pairs of integers given in
.I VALUE
against metadata. All non-digits are ignored in between.
.IP
.TS
tab(@)[];
lbl.
y<'show tracks before 2001.02.03. please'@Matches "2000", "2000-04.10", "2001X02".
y<=2001.02.03@Also matches "2001 02 03".
o~flac o>44@High-resolution FLAC files.
n?=3@Third tracks.
.TE
.
.TP
~
Test whether given PCRE
.RI ( VALUE )
matches metadata. (Default operator.)
.IP
.TS
tab(@)[];
lbl.
t[0-9]+cent@Same as "t~[0-9]cent".
ftcent@Same as "ftcent~" not "ft~cent".
t'oast for breakfast'@Can be omitted since "'" makes it clear.
'oast for breakfast'@Detto but without keys specified.
.TE
.
.IP
.I VALUE
is matched caseless unless it contains uppercase letter (smartcase).
.IP
.TS
tab(@)[];
lbl.
t~ear@Case-insensitive; matches "EAR", "eAR", "Ear" and "Heart".
t~Ear@Case-sensitive; matches "Ear" only.
.TE
.
.
.P
If
.I VALUE
is omitted, it is taken from the currently playing file. If metadatum has
has multiplicity (separated by ";"), only the first occurrence is used.
.IP
.TS
tab(@)[];
lbl.
a@Other tracks by the artist.
T@Tracks from the currently playing album.
y A T@A bit more specific.
A@Same as A~'Good' when currently playing file has A='Good;Bad;Ugly'.
.TE
.
.P
If a file has no metadata (either because not scanned yet or does not have at
all),
.IR KEY s
marked with "+" match
.I VALUE
against URL. (Because it is always known and URL may contain lot's of valuable
information.) This behavior can be prevented by "?" that requires
.I KEY
to be set.
.IP
.TS
tab(@)[];
lbl.
a~jimmy t~sunshine@Returns unscanned file named 'Jimmy - Sunshine.mp3'.
a?jimmy@Would not return the one above.
n?.@Use "." (match any) after "?" to test whether key is set.
!l?.@Scan metadata. Pause playback for better efficiency.
.TE
.
.P
Between expressions "!" (negate), "&" (and), "|" (or) operators can be used.
"&" is the default, so it may be omitted. "(", ")" can be used for grouping.
.IP
.TS
tab(@)[];
lb l.
!(g~rock y<2000)@All but rock before 2000.
!(g~rock & y<2000)@Same but "&" is explicitly written out.
!g~rock | (( (y>2000)| y=2000))@Same.
v~(bla)@Syntax error.
v~(bla )@Regex syntax error.
v~(b|a)@Regex syntax error + syntax error.
v~'(bla)'@Correct.
v~'(b|a)'@Correct.
(v'mix\)')@Matches version containing "mix)".
.TE
.
.SH SORTING
See "SEARCHING".
.P
.I FIRST-LINE
:=
.I EXPR
.br
.I EXPR
:= [
.I KEY
[ "=" ] [ "-" ] ]...
.
.TP
.B =
Perform numeric sort. If no numbers are found, fall back to general string comparsion.
.IP
.TS
tab(@)[];
lb l.
t=i=@Sort by track number THEN by index. "track-A" and "track-B" will be in correct order, though they do not contain any number.
ad=n=ti=@Sort by artists THEN by disc/track number THEN by title THEN by index (just to be stable sort).
.TE
.
.TP
\-
Sort in reverse order.
.IP
.TS
tab(@)[];
lb l.
m-@Latest added tracks.
y-@Latest tracks.
l-@Longest tracks on the top.
.TE
.
.SH ENVIRONMENT
.TP
.B EDITOR
Editor to use for text-editing operations.
.
.TP
.B MUCK_HOME
Configuration home.
.IP
Always set for children.
.
.SH EXAMPLES
.
Read files from arguments.
.IP
.EX
muck i-scream.flac http://online-screams.com:8008/radio ~/Music ~/Downloads
.EE
.
.P
Read playlist from standard input.
.IP
.EX
muck <<"PLAYLIST"
#EXTM3U
#EXTINF:title="foobar",
# This is my favourite song:
i-scream.flac
http://online-screams.com:8008/radio
/home/user/Music
/home/user/Downloads
PLAYLIST
.EE
.
.P
Start random playing matching files with decreased volume.
.IP
.EX
muck -q '~"b u g"' -e '80*r'
.EE
.
.P
Print errors.
.IP
.EX
muck 2>>/tmp/err
.EE
.
.P
An energy-efficient setup: Disable software volume and let ALSA talk directly
to hardware. When run in an XTerm-compatible terminal emulator screen is not
updated when unfocused.
.IP
.EX
muck -f alsa -o hw -a anull -e is
.EE
.
.P
Using together with
.BR alsaequal .
.IP
.EX
muck -f alsa -c pcm_f32le -a anull -e is
.EE
.
.SH "SEE ALSO"
.BR ffmpeg (1),
.BR ffplay (1),
.BR pcre2syntax (1)
.
.SH AUTHORS
Written by zsugabubus.
.
.SH LICENSE
GPLv3+


muck.c
#include <locale.h>
#include <unistd.h>

#include <libavcodec/avcodec.h>
#include <libavdevice/avdevice.h>
#include <libavfilter/avfilter.h>
#include <libavformat/avformat.h>

#include "config.h"
#include "env.h"
#include "expr.h"
#include "file.h"
#include "files.h"
#include "lua.h"
#include "player.h"
#include "playlist.h"
#include "tui.h"

#define FEATURES \
	/* xmacro(flag, name) */ \
	xmacro(WITH_ICU, "icu") \
	xmacro(WITH_LUA, "lua") \

#define FEATURE_HAVE0 "-"
#define FEATURE_HAVE1 "+"
#define FEATURE_HAVE(flag) FEATURE_HAVE##flag

#define FLAG0 "+"
#define FLAG1 "+"
#define FEATURE_STR(ICU)

static void
bye(void)
{
	tui_destroy();

	Error error;
	error_reset(&error);
	playlists_save(&error);

#if CONFIG_VALGRIND
# if WITH_LUA
	l_destroy();
# endif
	player_destroy();
	files_destroy();
	expr_global_uninit();
#endif
}

static void
open_cmdline(int argc, char *argv[], FileReadError *error)
{
	Playlist *master = playlist_alloc_master();
	if (!master)
		goto fail_enomem;

	if (!argc) {
		if (!isatty(STDIN_FILENO)) {
			File *f = playlist_alloc_file_dupurl(master, F_PLAYLIST, "stdin");
			if (!f)
				goto fail_enomem;
			Playlist *playlist = playlist_alloc(f, f->url);
			if (!playlist)
				goto fail_enomem;
			playlist->read_only = 1;
			playlist->dirname = strdup(".");
			if (!playlist->dirname)
				goto fail_enomem;
			playlist->dirfd = dup(master->dirfd);
			playlist_read_m3u(playlist, dup(STDIN_FILENO), error);
		} else {
			File *f = playlist_alloc_file_dupurl(master, F_PLAYLIST_DIRECTORY, ".");
			if (!f)
				goto fail_enomem;
			file_read(f, error);
		}
	} else for (int i = 0; i < argc; ++i) {
		char const *url = argv[i];
		enum FileType type = playlist_probe_url(master, url);
		File *f = playlist_alloc_file_dupurl(master, type, url);
		if (!f)
			goto fail_enomem;
		file_read(f, error);
	}

	return;

fail_enomem:
	error_from_strerror(&error->error, ENOMEM);
}

int
main(int argc, char *argv[])
{
	setlocale(LC_ALL, "");
	atexit(bye);

	Error error;
	error_reset(&error);

	env_init(&error);
	expr_global_init(&error);
	files_init(&error);
	tui_init();
	player_init(&error);

	error_ok_or_die(&error, "Could not initalize globals");

	char const *startup_cmd = "0G";
	char const *codec = "pcm";
	char const *format_name = "alsa";
	char const *filename = NULL;
	char const *graph_descr = "volume=replaygain=track";

	for (int c; 0 <= (c = getopt(argc, argv, "q:e:a:c:f:o:m:C:s:dv"));)
		switch (c) {
		case 'q':
		{
			ExprParserContext parser;
			error_reset(&parser.error);

			char *s = strdup(optarg);
			if (!s)
				error_from_errno(&parser.error);
			else
				files_set_filter(&parser, s);
			error_ok_or_die(&parser.error, "invalid argument for 'q'");
		}
			break;

		case 'e':
			startup_cmd = optarg;
			break;

		case 'a':
			graph_descr = optarg;
			break;

		case 'c':
			codec = optarg;
			break;

		case 'f':
			format_name = optarg;
			break;

		case 'o':
			filename = optarg;
			break;

		case 'm':
			player_set_buffer(strtoll(optarg, NULL, 10) * 1024);
			break;

		case 'C':
		{
			char *s = strdup(optarg);
			if (!s)
				error_from_errno(&error);
			error_ok_or_die(&error, "invalid argument for 'C'");
			tui_set_columns(s);
		}
			break;

		case 's':
		{
			char *s = strdup(optarg);
			if (!s)
				error_from_errno(&error);
			error_ok_or_die(&error, "invalid argument for 's'");
			files_set_order(s);
		}
			break;

		case 'd':
			av_log_set_level(av_log_get_level() < AV_LOG_DEBUG ? AV_LOG_DEBUG : AV_LOG_TRACE);
			break;

		case 'v':
#define xmacro(flag, name) " " FEATURE_HAVE(flag) name
			puts(MUCK_VERSION);
			puts("Features:" FEATURES);
#undef xmacro
			printf("FFmpeg %s:\n"
					"  %s: %s\n"
					"  %s: %s\n"
					"  %s: %s\n"
					"  %s: %s\n"
					"  %s: %s\n",
					av_version_info(),
					LIBAVUTIL_IDENT, avutil_configuration(),
					LIBAVCODEC_IDENT, avcodec_configuration(),
					LIBAVFORMAT_IDENT, avformat_configuration(),
					LIBAVFILTER_IDENT, avfilter_configuration(),
					LIBAVDEVICE_IDENT, avdevice_configuration());
			return EXIT_SUCCESS;

		default:
			return EXIT_FAILURE;
		}

#if WITH_LUA
	l_open();
#endif

	player_configure(format_name, filename, codec, graph_descr);

	player_run(&error);
	error_ok_or_die(&error, "Could not start player");

	FileReadError read_error;
	error_reset(&read_error.error);
	open_cmdline(argc - optind, argv + optind, &read_error);
	error_ok_or_die(&read_error.error, "Could not populate files from command line");

	tui_feed_keys(startup_cmd);
	tui_run();
}
